buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -229,18 +229,111 @@ public static int hashCode(ByteBuf buffer) {
-        for (int i = 0; i < attempts; i++) {
-            if (equals(needle, needle.readerIndex(),
-                       haystack, haystack.readerIndex() + i,
-                       needle.readableBytes())) {
-                return haystack.readerIndex() + i;
+        if (haystack == null || needle == null) {
+            return -1;
+        }
+
+        if (needle.readableBytes() > haystack.readableBytes()) {
+            return -1;
+        }
+
+        int n = haystack.readableBytes();
+        int m = needle.readableBytes();
+        if (m == 0) {
+            return 0;
+        }
+
+        int i = maxSuf(needle, m, true);
+        int j = maxSuf(needle, m, false);
+        int ell = Math.max(i, j);
+        int memory;
+        int per = 0;
+
+        if (equals(needle, 0, needle, per, per + ell + 1)) {
+            j = 0;
+            memory = -1;
+            while (j <= n - m) {
+                i = Math.max(ell, memory) + 1;
+                while (i < m && needle.getByte(i) == haystack.getByte(i + j)) {
+                    ++i;
+                }
+                if (i > n) {
+                    return -1;
+                }
+                if (i >= m) {
+                    i = ell;
+                    while (i > memory && needle.getByte(i) == haystack.getByte(i + j)) {
+                        --i;
+                    }
+                    if (i <= memory) {
+                        return j;
+                    }
+                    j += per;
+                    memory = m - per - 1;
+                } else {
+                    j += i - ell;
+                    memory = -1;
+                }
+            }
+        } else {
+            per = Math.max(ell + 1, m - ell - 1) + 1;
+            j = 0;
+            while (j <= n - m) {
+                i = ell + 1;
+                while (i < m && needle.getByte(i) == haystack.getByte(i + j)) {
+                    ++i;
+                }
+                if (i > n) {
+                    return -1;
+                }
+                if (i >= m) {
+                    i = ell;
+                    while (i >= 0 && needle.getByte(i) == haystack.getByte(i + j)) {
+                        --i;
+                    }
+                    if (i < 0) {
+                        return j;
+                    }
+                    j += per;
+                } else {
+                    j += i - ell;
+                }
             }
         }
         return -1;
     }
 
+    private static int maxSuf(ByteBuf x, int m, boolean isSuffix) {
+        int p = 1;
+        int ms = -1;
+        int j = 0;
+        int k = 1;
+        byte a;
+        byte b;
+        while (j + k < m) {
+            a = x.getByte(j + k);
+            b = x.getByte(ms + k);
+            boolean suffix = isSuffix ? a < b : a > b;
+            if (suffix) {
+                j += k;
+                k = 1;
+                p = j - ms;
+            } else if (a == b) {
```

Comment:
Please add some tests that exercises this branch, where `a` and `b` are equal.

Reply:
I added a test case, when the first byte of needle is equal to the second byte, a will be equal to b